### DISCUSSION TITLE
Remove `KeyedColumnElement` since it seems unnecessary

### DISCUSSION THREAD

**Discussion by CaselIT** [sqlalchemy/sqlalchemy#10325] [Ideas]
Currently InstrumentedAttribute is not a KeyedColumnElement so trying to use `corresponding_column` and similar raises a typing error

This does not seem correct since at runtime it it seems to work
```py
from typing import TYPE_CHECKING

from sqlalchemy import Select
from sqlalchemy.orm import column_property
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column

class Base(DeclarativeBase):
        pass

class Widget(Base):
    __tablename__ = "widget"
    id: Mapped[int] = mapped_column(primary_key=True)
    data: Mapped[int]

    if TYPE_CHECKING:
        foo: Mapped[int]

Widget.foo = column_property((Widget.data + 1).label('foo'))

q = select(Widget)
print(q)
s = q.subquery()
print(s.columns.corresponding_column(Widget.foo))
Widget.data.key = None
print(Widget.data.key, s.columns.corresponding_column(Widget.data))
```

the above prints
```
SELECT widget.id, widget.data, widget.data + :data_1 AS foo
FROM widget
anon_1.foo
None anon_1.data
```

given that it works also in the last case, what's the point of `KeyedColumnElement`?

**Comment by zzzeek** [comment]
`KeyedColumnElement` is a `ColumnElement` that has a string `.key` attribute.  `InstrumentedAttribute` objects are not `ColumnElement` objects.  

it's technically not correect to send `InstrumentedAttribute` to `.corresponding_column()` right now since the latter does not use coercion, it just "happens" to work by accident right now.    `.corresponding_column()` can be improved to formally accept objects with coercsion added, in which case it would accept something like `roles.ColumnArgumentRole`, something very basic like that.

**Reply by zzzeek** [reply]
of coruse if we add coercion to `.corresponding_column()`, all the internals should be switched to use a private version that does not have this performance overhead since this method is fairly performance critical

**Reply by CaselIT** [reply]
ok. I'll open an issue then

**Reply by CaselIT** [reply]
opened #10326
